### PR TITLE
[None][fix] preserve multimodal UUID in V2 KV events

### PR DESCRIPTION
**Pull request by GuanLuo**
## Description

Buffered KV cache event protocol V2 currently derives multimodal tokens from content digests, but drops the frontend-provided multimodal UUID before events are serialized. This causes routing observers to see a different identity from the one attached to the request.

This change carries both the digest and optional multimodal UUID through the native and Python V2 token/radix/event paths. Digest equality and hashing remain unchanged for cache lookup and reuse, while serialized V2 events now add `uuid` when one was provided. UUID-less inputs keep the existing schema, legacy tuple forms remain accepted, and protocol V1 retains its existing UUID-as-hash behavior. The request binding, documentation, and regression coverage are updated accordingly.

Functional impact is limited to preserving multimodal identity in V2 buffered event payloads; no cache-key or cache-reuse semantics change. No performance regression is expected.

## Test Coverage

- All pre-commit hooks passed for changed files.
- Native V2 digest-pool and event tests: 8 passed.
- Buffered/API Python regressions: 99 passed, 1 skipped.
- Focused Python backend identity parity tests: 5 passed.
- Targeted GPU multimodal V1/V2 tests: 6 passed.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

## Dev Engineer Review

The change preserves multimodal UUIDs through native and Python V2 token, radix-tree, and event paths. V2 events retain the digest in `hash` and emit `uuid` when supplied. UUID-less inputs retain the existing schema.

Digest equality, hashing, lookup, and cache reuse remain digest-based. Legacy tuple forms remain supported. V1 retains its UUID-as-`hash` behavior. The main review risk is consistency across C++, Python bindings, type stubs, and serializers.

## QA Engineer Review

Modified tests cover UUID propagation, digest-based identity, cache reuse, partial UUID lists, V1/V2 serialization, block ancestry, lifecycle removal, detached ancestors, and long UUID values.

Applicable test-list entries include:
- `tests/integration/test_lists/test-db/l0_b200.yml`
  - `test_mm_encoder_standalone.py -k "qwen3_30b_a3b_fp8"`
  - `test_mm_encoder_standalone.py -k "test_kv_event_mm_keys and uuid"`
- `tests/integration/test_lists/test-db/l0_dgx_h200.yml`
  - `test_llm_kv_cache_events.py::test_llm_api_attention_dp_kv_events`

Two existing manual waivers cover selected standalone multimodal tests. Test execution results are not supplied. Coverage verdict: **sufficient**.

## Per-File QA Perspective

- `cpp/tensorrt_llm/batch_manager/kv_cache_manager_v2/blockRadixTree.cpp`: Verify UUID-bearing digest tokens and context propagation.
- `cpp/tensorrt_llm/batch_manager/kv_cache_manager_v2/blockRadixTree.h`: Verify the optional UUID API and context accessor.
- `cpp/tensorrt_llm/batch_manager/kv_cache_manager_v2/eventManager.cpp`: Verify digest, offset, and UUID preservation across blocks.
- `cpp/tensorrt_llm/batch_manager/kv_cache_manager_v2/eventManager.h`: Verify `uuidIsAdditive` affects key equality and serialization.
- `cpp/tensorrt_llm/batch_manager/kv_cache_manager_v2/tokenIdExt.cpp`: Verify pooled context allocation and reclamation.
- `cpp/tensorrt_llm/batch_manager/kv_cache_manager_v2/tokenIdExt.h`: Verify context construction and digest-only equality.
- `cpp/tensorrt_llm/nanobind/batch_manager/bindings.cpp`: Verify `multimodal_uuids` returns the expected optional values.
- `cpp/tensorrt_llm/nanobind/batch_manager/kvCacheManagerV2.cpp`: Verify Python construction, hashing, pickle support, tuple compatibility, and UUID forwarding.
- `cpp/tests/unit_tests/batch_manager/kvCacheManagerV2DigestPoolTest.cpp`: Covers UUID metadata, copy/move behavior, generated tokens, and unchanged digest identity. Listed CI coverage was not confirmed.
- `docs/source/features/kvcache.md`: Verify documentation matches V1/V2 event fields.
- `tensorrt_llm/_torch/pyexecutor/kv_cache/kv_cache_manager_v2.py`: Verify UUID selection for contiguous and run-based inputs, including missing entries.
- `tensorrt_llm/_torch/pyexecutor/kv_cache_events.py`: Verify `bytes` and `MmItemContext` tokens follow the multimodal path.
- `tensorrt_llm/_utils.py`: Verify legacy and four-element key serialization.
- `tensorrt_llm/inputs/data.py`: Verify prompt documentation matches runtime behavior.
- `tensorrt_llm/inputs/multimodal.py`: Verify public UUID documentation matches event output.
- `tensorrt_llm/inputs/registry.py`: Verify hashing documentation matches UUID behavior.
- `tensorrt_llm/runtime/kv_cache_manager_v2/__init__.py`: Verify consistent `MmItemContext` exports.
- `tensorrt_llm/runtime/kv_cache_manager_v2/__init__.pyi`: Verify stubs match runtime APIs.
- `tensorrt_llm/runtime/kv_cache_manager_v2/_block_radix_tree.py`: Verify context propagation and digest-only hashing.
- `tensorrt_llm/runtime/kv_cache_manager_v2/_common.py`: Verify context equality and hashing ignore UUID metadata.
- `tensorrt_llm/runtime/kv_cache_manager_v2/_core/_kv_cache.py`: Verify text-only mode rejects `MmItemContext`.
- `tensorrt_llm/runtime/kv_cache_manager_v2/_event_manager.py`: Verify context inheritance and four-element key generation.
- `tests/unittest/_torch/executor/kv_cache/test_kv_cache_v2_multimodal_runs.py`: Covers reuse identity, UUID preservation, and partial UUID inputs. CI listing was not found.
- `tests/unittest/_torch/multimodal/test_mm_encoder_standalone.py`: Covers V1/V2 UUID serialization and edge cases. Listed in `l0_b200.yml`; selected tests also have waivers.
- `tests/unittest/kv_cache_manager_v2_tests/kernels.py`: Verifies token conversion for `MmItemContext`.
- `tests/unittest/kv_cache_manager_v2_tests/test_kv_cache_event_manager.py`: Covers UUID metadata across ancestry and block lifecycle. CI listing was not found.
- `tests/unittest/llmapi/test_llm_kv_cache_events.py`: Covers V2 `hash`/`uuid` separation. Listed in `l0_dgx_h200.yml`.

### CODE REVIEW

**Review by SimengLiu-nv — APPROVED** [review]
Functional changes lgtm.

**Review by Shixiaowei02 — APPROVED** [review]
Approve from a documentation perspective. Thanks!

File: cpp/tensorrt_llm/batch_manager/kv_cache_manager_v2/tokenIdExt.cpp [hunk 1]
```diff
@@ -74,28 +76,28 @@ class DigestPool
...
     // released and across later shrinks (which only pop free tail slots). Uses
     // at() so a bad index (e.g. the sentinel of a moved-from handle) throws
-    // rather than reading out of bounds; digests are rare so the check is cheap.
-    [[nodiscard]] Digest const& get(uint32_t idx) const
+    // rather than reading out of bounds; contexts are rare so the check is cheap.
+    [[nodiscard]] MmItemContext const& get(uint32_t idx) const
     {
         std::lock_guard<std::mutex> const lock(mMutex);
-        return mStore.at(idx);
+        return *mStore.at(idx);
```
Comment (brnguyen2) [inline cpp/tensorrt_llm/batch_manager/kv_cache_manager_v2/tokenIdExt.cpp]: `free()` now resets the slot's `shared_ptr`, so an in-range but freed index reaches here as a null pointer and `*mStore.at(idx)` is undefined behavior rather than the `std::out_of_range` the comment above promises. Before this change the same stale index returned a stale-but-valid `Digest`. `at()` only protects the out-of-range sentinel case (which the tests cover). Suggest guarding the slot explicitly, e.g.

```cpp
auto const& slot = mStore.at(idx);
TLLM_CHECK_WITH_INFO(slot != nullptr, "DigestPool slot %u is not live", idx);
return *slot;
```

Digest tokens are rare, so the check is free, and it turns a latent use-after-free class of bug into a hard error.

File: cpp/tensorrt_llm/batch_manager/kv_cache_manager_v2/blockRadixTree.cpp [hunk 2]
```diff
@@ -330,13 +330,13 @@ Block::Block(BlockKey k, std::vector<TokenIdExt> toks, NodeBase* prevNode)
...
         {
-            mLastTokenDigest = static_cast<Block const*>(prevNode)->getLastTokenDigest();
+            mLastMmItemContext = static_cast<Block const*>(prevNode)->getLastMmItemContext();
         }
         for (auto iter = tokens.rbegin(); iter != tokens.rend(); ++iter)
         {
             if (iter->isDigest())
             {
-                mLastTokenDigest = std::make_shared<Digest const>(iter->digest());
+                mLastMmItemContext = std::make_shared<MmItemContext const>(iter->mmItemContext());
```
Comment (brnguyen2) [inline cpp/tensorrt_llm/batch_manager/kv_cache_manager_v2/blockRadixTree.cpp]: This deep-copies the `MmItemContext` (including the UUID string, which the tests show can be kilobytes) once per block boundary that follows a digest token, even though the pool already holds this exact context behind a `shared_ptr` and `duplicate()` shares it for token copies. Optional: expose the pool's `shared_ptr<MmItemContext const>` (e.g. `TokenIdExt::sharedMmItemContext()`) and store that here instead of `make_shared` of a copy. Not a correctness issue, just avoids the string copy the shared-pointer redesign was meant to eliminate.

File: cpp/tensorrt_llm/batch_manager/kv_cache_manager_v2/eventManager.h [hunk 3]
```diff
@@ -70,11 +70,12 @@ struct MmKey
     int startOffset = 0;
     std::optional<std::string> uuid;
     bool hasUuidField = false;
+    bool uuidIsAdditive = false;
```
Comment (brnguyen2) [inline cpp/tensorrt_llm/batch_manager/kv_cache_manager_v2/eventManager.h]: On the V2 event path `hasUuidField` and `uuidIsAdditive` are always set to the same value (`uuid.has_value()`), so `MmKey` now has three fields encoding two possible output shapes, and a future caller can construct the contradictory `{uuid=nullopt, hasUuidField=false, uuidIsAdditive=true}`. The legacy 3-tuple shape only arises from Python-supplied tuples in `castMmKeys(nb::handle)`. Optional cleanup: replace both booleans with a small `enum class UuidMode { kNone, kReplacesHash, kAdditive }` so the wire shape is a single unambiguous value and `castMmKeys` becomes a switch.

File: tensorrt_llm/_torch/pyexecutor/kv_cache/kv_cache_manager_v2.py [hunk 4]
```diff
@@ -4217,13 +4234,20 @@ def _augment_tokens_for_block_reuse(
         run_metadata = _resolve_multimodal_run_metadata(req)
         if run_metadata is not None:
             return _augment_tokens_with_mm_run_metadata(
-                self.vocab_size, result, req.multimodal_hashes, run_metadata, chunk_start, chunk_end
+                self.vocab_size,
+                result,
+                req.multimodal_hashes,
+                getattr(req, "multimodal_uuids", None),
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/pyexecutor/kv_cache/kv_cache_manager_v2.py]: `req.multimodal_hashes` two lines above is read directly, but `multimodal_uuids` goes through `getattr(..., None)` here and at the second call site below. This PR adds the `multimodal_uuids` property to the `LlmRequest` binding, so the fallback can only trigger if that binding is missing, in which case UUIDs are silently dropped from events, which is exactly the bug this PR fixes. Suggest `req.multimodal_uuids` so a missing binding fails loudly, matching how the sibling fields are accessed.

**Review by brnguyen2 — APPROVED** [review]
Approving — the comments below are optional touch-ups, not blockers.

The digest/UUID split is enforced at every layer that can observe a token (C++ pool, Python dataclass, nanobind `__eq__`/`__hash__`, both hashers), and the event-manager parity test now runs the same UUID scenario against both backends, so the main consistency risk called out in the description is covered. I checked for consumers that destructure the raw `mm_keys` tuples outside the JSON serializer and for `api_stability` references to the widened `TokenIdExt` / new `uuid` kwarg; both are clean, so the 4-tuple form stays internal.

Two whole-PR notes:

- The PR is two commits (`preserve multimodal UUID in V2 KV events` plus `align native multimodal context hashing`). Please squash before merge, or make sure the first commit builds and passes on its own; a bisect landing between them would otherwise hit a C++/Python hashing mismatch.
- `docs/source/features/kvcache.md` is the only doc that describes the `mm_keys` schema and it is updated, including the note that `uuid` is the only way to recover the external identity. Good. The `TextPrompt`/`TokensPrompt`/`MultimodalInput` docstrings now match V1/V2 behavior too.

The inline comments are optional hardening and cleanups; none block merge.
